src/OpenTelemetry.Contrib.Extensions.AWSXRay/Instrumentation/AWSTracingPipelineHandler.cs

```diff
@@ -0,0 +1,249 @@
+        };
+
+        private static readonly ActivitySource AWSSDKActivitySource = new ActivitySource(ActivitySourceName);
+
+        public override void InvokeSync(IExecutionContext executionContext)
+        {
+            var activity = this.ProcessBeginRequest(executionContext);
+            try
+            {
+                base.InvokeSync(executionContext);
+            }
+            catch (Exception ex)
+            {
+                if (activity != null)
+                {
+                    this.ProcessException(activity, ex);
+                }
+
+                throw;
+            }
+            finally
+            {
+                if (activity != null)
+                {
+                    this.ProcessEndRequest(executionContext, activity);
+                }
+            }
+        }
+
+        public override async Task<T> InvokeAsync<T>(IExecutionContext executionContext)
+        {
+            T ret = null;
+
+            var activity = this.ProcessBeginRequest(executionContext);
+            try
+            {
+                ret = await base.InvokeAsync<T>(executionContext).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                if (activity != null)
+                {
+                    this.ProcessException(activity, ex);
+                }
+
+                throw;
+            }
+            finally
+            {
+                if (activity != null)
+                {
+                    this.ProcessEndRequest(executionContext, activity);
+                }
+            }
+
+            return ret;
+        }
+
+        private Activity ProcessBeginRequest(IExecutionContext executionContext)
+        {
+            Activity activity = null;
+
+            var requestContext = executionContext.RequestContext;
+            var service = this.GetAWSServiceName(requestContext);
+            var operation = this.GetAWSOperationName(requestContext);
+
+            activity = AWSSDKActivitySource.StartActivity(service + "." + operation, ActivityKind.Client);
+
+            if (activity == null)
+            {
+                return null;
+            }
+
+            activity.SetTag(AWSSemanticConventions.AttributeAWSServiceName, service);
+            activity.SetTag(AWSSemanticConventions.AttributeAWSOperationName, operation);
+            var client = executionContext.RequestContext.ClientConfig;
+            if (client != null)
+            {
+                var region = client.RegionEndpoint?.SystemName;
+                activity.SetTag(AWSSemanticConventions.AttributeAWSRegion, region ?? AWSSDKUtils.DetermineRegion(client.ServiceURL));
+            }
+
+            this.AddRequestSpecificInformation(activity, requestContext, service);
+
+            AwsPropagator.Inject(new PropagationContext(activity.Context, Baggage.Current), requestContext.Request.Headers, Setter);
+
+            return activity;
```

Comment:
I was expecting to see a check in here like...
```csharp
            if (this.options.SuppressDownstreamInstrumentation)
            {
                SuppressInstrumentationScope.Enter();
            }
```
That would prevent an http span from also being created if user also has http instrumentation turned on. Not sure if this was left out intentionally, just thought I'd bring it up.

Reply:
Currently I'm leveraging [this](https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs#L86-L91) piece of logic in http instrumentation to suppress the downstream instrumentation by injecting the context in the AWSXRayPropagator. The catch is that if the user also has http instrumentation, he/she would need to use the AWSXRayPropagator in the instrumentation.

This might be a better approach (which I didn't know about) to suppressing the http instrumentation. :) Thanks!

Reply:
@srprash The code there you linked is really meant for redirects and ssl negotation in the .net http clients where the request is re-used. `SuppressInstrumentationScope` is the official mechanism provided by the API/SDK. Probably better to use that because it will always be supported. Here's an example of usage:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/da4d352a198a89200fcec74bd4a20838aec45e02/src/OpenTelemetry.Instrumentation.GrpcNetClient/Implementation/GrpcClientDiagnosticListener.cs#L50-L53

There's also an example in there of `activity.IsAllDataRequested` being checked.

(GrpcNetClient instrumentation is a good example for you to use for this because it is also an instrumentation layer built on top of HttpClient.)

Reply:
Thanks @CodeBlanch ! I've changed the instrumentation to use the `SuppressInstrumentationScope` now and it works! :)